// filestack-js/build/main/lib/request/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FsRequestError = exports.FsRequestErrorCode = void 0;
var tslib_1 = require("tslib");
/**
 * Filestack error codes (common for browser and node)
 *
 * @export
 * @enum {number}
 */
var FsRequestErrorCode;
(function (FsRequestErrorCode) {
    FsRequestErrorCode["ABORTED"] = "ABORTED";
    FsRequestErrorCode["NETWORK"] = "NETWORK";
    FsRequestErrorCode["TIMEOUT"] = "TIMEDOUT";
    FsRequestErrorCode["SERVER"] = "SERVER";
    FsRequestErrorCode["REQUEST"] = "REQUEST";
    FsRequestErrorCode["OTHER"] = "OTHER";
    FsRequestErrorCode["REDIRECT"] = "REDIRECT";
})(FsRequestErrorCode = exports.FsRequestErrorCode || (exports.FsRequestErrorCode = {}));
/**
 * Custom Filestack Request error class
 *
 * @export
 * @class FsRequestError
 * @extends {Error}
 */
var FsRequestError = /** @class */ (function (_super) {
    tslib_1.__extends(FsRequestError, _super);
    /**
     * Creates an instance of FsRequestError.
     *
     * @param {string} message
     * @param {*} config
     * @param {FsResponse} [response]
     * @param {FsRequestErrorCode} [code]
     * @memberof FsRequestError
     */
    function FsRequestError(message, config, response, code) {
        var _newTarget = this.constructor;
        var _this = 
        /* istanbul ignore next */
        _super.call(this, message) || this;
        _this.config = config;
        _this.response = response;
        _this.code = code;
        // const captureStackTrace: Function = (Error as any).captureStackTrace;
        // captureStackTrace && captureStackTrace(this);
        fixProto(_this, _newTarget.prototype);
        return _this;
    }
    return FsRequestError;
}(Error));
exports.FsRequestError = FsRequestError;
function fixProto(target, prototype) {
    var setPrototypeOf = Object.setPrototypeOf;
    /* istanbul ignore next */
    setPrototypeOf ? setPrototypeOf(target, prototype) : (target.__proto__ = prototype);
}
