// filestack-js/build/main/lib/utils/store.browser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Store = void 0;
var tslib_1 = require("tslib");
/*
 * Copyright (c) 2018 by Filestack.
 * Some rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var debug_1 = tslib_1.__importDefault(require("debug"));
var store_1 = require("./store");
var debug = (0, debug_1.default)('fs:storage');
var Store = /** @class */ (function () {
    function Store() {
        this.availableTypes = [];
        if (window.localStorage !== undefined) {
            this.availableTypes.push(store_1.STORE_TYPE.LOCAL);
            debug('Local storage exists');
        }
        if (window.sessionStorage !== undefined) {
            this.availableTypes.push(store_1.STORE_TYPE.SESSION);
            debug('Session storage exists');
        }
        debug('availableTypes - %O', this.availableTypes);
    }
    Store.prototype.hasAccess = function (prefType) {
        debug('Checking access to storage for %s. Available types: %O', prefType, this.availableTypes);
        if (prefType) {
            return this.availableTypes.indexOf(prefType) > -1;
        }
        return !!this.availableTypes.length;
    };
    Store.prototype.getItem = function (key, prefType) {
        var store = this.getStore(prefType);
        debug('getItem %s from %O', key, store);
        if (!store) {
            return undefined;
        }
        return store.getItem(key);
    };
    Store.prototype.setItem = function (key, value, prefType) {
        var store = this.getStore(prefType);
        debug('setItem %s to %O', key, store);
        if (!store) {
            return undefined;
        }
        return store.setItem(key, value);
    };
    Store.prototype.removeItem = function (key, prefType) {
        var store = this.getStore(prefType);
        debug('removeItem %s from %O', key, store);
        if (!store) {
            return undefined;
        }
        return store.removeItem(key);
    };
    Store.prototype.getStore = function (prefType) {
        debug('get store with pref type - %s', prefType);
        if (!this.hasAccess(prefType)) {
            debug('User dont have access to any storage');
            return undefined;
        }
        switch (prefType) {
            case store_1.STORE_TYPE.LOCAL:
                return window.localStorage;
            case store_1.STORE_TYPE.SESSION:
                return window.sessionStorage;
            default:
                // @ts-ignore
                return window[this.availableTypes[0]];
        }
    };
    return Store;
}());
exports.Store = Store;
