// filestack-js/build/main/lib/request/helpers/shouldRetry.spec.js
"use strict";
/*
 * Copyright (c) 2018 by Filestack
 * Some rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var shouldRetry_1 = require("./shouldRetry");
var error_1 = require("../error");
var testResponse = function (code) {
    if (code === void 0) { code = 200; }
    return ({
        status: code,
        statusText: 'test',
        headers: {},
        data: {},
        config: {},
    });
};
describe('Request/Helpers/shouldRetry', function () {
    describe('shouldRetry', function () {
        it('should retry on network error', function () {
            expect((0, shouldRetry_1.shouldRetry)(new error_1.FsRequestError('', null, testResponse(), error_1.FsRequestErrorCode.NETWORK))).toEqual(true);
        });
        it('should retry on server error', function () {
            expect((0, shouldRetry_1.shouldRetry)(new error_1.FsRequestError('', null, testResponse(), error_1.FsRequestErrorCode.SERVER))).toEqual(true);
        });
        it('should retry on server error', function () {
            expect((0, shouldRetry_1.shouldRetry)(new error_1.FsRequestError('', null, testResponse(), error_1.FsRequestErrorCode.TIMEOUT))).toEqual(true);
        });
        it('should retry on request 5xx code', function () {
            // @ts-ignore
            expect((0, shouldRetry_1.shouldRetry)(new error_1.FsRequestError('', null, testResponse(504), error_1.FsRequestErrorCode.OTHER))).toEqual(true);
        });
        it('should not retry on 4xx codes', function () {
            // @ts-ignore
            expect((0, shouldRetry_1.shouldRetry)(new error_1.FsRequestError('', null, testResponse(404), error_1.FsRequestErrorCode.REQUEST))).toEqual(false);
        });
        it('should not retry on request error', function () {
            // @ts-ignore
            expect((0, shouldRetry_1.shouldRetry)(new error_1.FsRequestError('', null, testResponse(), error_1.FsRequestErrorCode.REQUEST))).toEqual(false);
        });
        it('should not retry on request aborted', function () {
            // @ts-ignore
            expect((0, shouldRetry_1.shouldRetry)(new error_1.FsRequestError('', null, testResponse(), error_1.FsRequestErrorCode.ABORTED))).toEqual(false);
        });
    });
});
