// filestack-js/build/main/lib/api/transform.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/*
 * Copyright (c) 2018 by Filestack.
 * Some rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var transform_1 = require("./transform");
var config_1 = require("./../../config");
var filelink_1 = require("./../filelink");
jest.mock('./../filelink');
var defaultApikey = 'EXAMPLE_API_KEY';
var defaultHandle = 'EXAMPLE_HANDLE';
var defaultSecurity = {
    policy: 'examplePolicy',
    signature: 'exampleSignature',
};
var sessionURls = config_1.config.urls;
var defaultSession = {
    apikey: defaultApikey,
    urls: sessionURls,
};
describe('OldTransforms', function () {
    it('should pass params to Filelink class', function () {
        (0, transform_1.transform)(defaultSession, defaultHandle, {
            partial_pixelate: {
                amount: 2,
            },
        });
        expect(filelink_1.Filelink.prototype.addTask).toHaveBeenCalledWith('partial_pixelate', {
            amount: 2,
        });
    });
    it('should respect security params', function () {
        (0, transform_1.transform)(tslib_1.__assign(tslib_1.__assign({}, defaultSession), defaultSecurity), defaultHandle, {
            partial_pixelate: {
                amount: 2,
            },
        });
        expect(filelink_1.Filelink.prototype.addTask).toHaveBeenCalledWith('security', defaultSecurity);
    });
    it('should respect cache=false transformation', function () {
        (0, transform_1.transform)(defaultSession, defaultHandle, {
            // @ts-ignore
            cache: false,
        });
        expect(filelink_1.Filelink.prototype.addTask).toHaveBeenCalledWith('cache', false);
    });
    it('should remove falsy parameters transformation', function () {
        (0, transform_1.transform)(defaultSession, defaultHandle, {
            flip: false,
        });
        expect(filelink_1.Filelink.prototype.addTask).not.toHaveBeenCalledWith('flip', false);
    });
    it('should change cammel case to snake case transformations', function () {
        (0, transform_1.transform)(defaultSession, defaultHandle, {
            // @ts-ignore
            partialPixelate: {
                amount: 2,
            },
        });
        expect(filelink_1.Filelink.prototype.addTask).toHaveBeenCalledWith('partial_pixelate', {
            amount: 2,
        });
    });
    it('return call toString on filelink when params are empty', function () {
        var testUrl = 'nanana';
        jest.spyOn(filelink_1.Filelink.prototype, 'toString').mockImplementation(function () { return testUrl; });
        expect((0, transform_1.transform)(defaultSession, defaultHandle)).toEqual(testUrl);
    });
    it('should enable base64 on filelink', function () {
        (0, transform_1.transform)(defaultSession, defaultHandle, {
            // @ts-ignore
            partialPixelate: {
                amount: 2,
            },
        }, true);
        expect(filelink_1.Filelink.prototype.setBase64).toHaveBeenCalledWith(true);
    });
});
