// filestack-js/build/main/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrefetchEvents = exports.getMimetype = exports.PickerDisplayMode = exports.FilestackError = exports.Client = exports.validateWebhookSignature = exports.getSecurity = exports.TransformSchema = exports.version = exports.init = exports.Filestack = void 0;
var tslib_1 = require("tslib");
/*
 * Copyright (c) 2018 by Filestack.
 * Some rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var loader_1 = require("@filestack/loader");
var security_1 = require("./lib/api/security");
Object.defineProperty(exports, "getSecurity", { enumerable: true, get: function () { return security_1.getSecurity; } });
Object.defineProperty(exports, "validateWebhookSignature", { enumerable: true, get: function () { return security_1.validateWebhookSignature; } });
var client_1 = require("./lib/client");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return client_1.Client; } });
var picker_1 = require("./lib/picker");
Object.defineProperty(exports, "PickerDisplayMode", { enumerable: true, get: function () { return picker_1.PickerDisplayMode; } });
var transforms_schema_1 = require("./schema/transforms.schema");
Object.defineProperty(exports, "TransformSchema", { enumerable: true, get: function () { return transforms_schema_1.TransformSchema; } });
var prefetch_1 = require("./lib/api/prefetch");
Object.defineProperty(exports, "PrefetchEvents", { enumerable: true, get: function () { return prefetch_1.PrefetchEvents; } });
var filestack_error_1 = require("./filestack_error");
Object.defineProperty(exports, "FilestackError", { enumerable: true, get: function () { return filestack_error_1.FilestackError; } });
var index_1 = require("./lib/utils/index");
Object.defineProperty(exports, "getMimetype", { enumerable: true, get: function () { return index_1.getMimetype; } });
/**
 * Initialize client with given config
 *
 * @param apikey
 * @param options
 */
var Filestack = function (apikey, options) {
    return new client_1.Client(apikey, options);
};
exports.Filestack = Filestack;
// This will be deprecated in feature use
exports.init = exports.Filestack;
/**
 * filestack-js version. Interpolated at build time.
 */
exports.version = '3.36.4';
(0, loader_1.registerModule)(loader_1.FILESTACK_MODULES.FILESTACK_SDK, exports.Filestack, { version: '3.36.4' });
tslib_1.__exportStar(require("./lib/api/transform"), exports);
tslib_1.__exportStar(require("./lib/filelink"), exports);
tslib_1.__exportStar(require("./filestack_error"), exports);
tslib_1.__exportStar(require("./lib/request"), exports);
tslib_1.__exportStar(require("./lib/utils"), exports);
